### ISSUE TITLE
File is not parsed with C++17 support

### ISSUE THREAD

**Issue by jmigual** [clangd/clangd#823]
I have a project where some files are compiled with C++17 while others aren't (they are external libraries). When parsing headers inside a nested folder, clangd says that `std::optional` does not exist but this is a header file only used by files compiled with C++17.

I've provided a minimal working example here: [cpp-experiments.zip](https://github.com/clangd/clangd/files/6845990/cpp-experiments.zip)

In the example, in the file `include/math/my_header.hpp`, clangd says that optional does not exist:
![imatge](https://user-images.githubusercontent.com/6493651/126273525-ade659e4-43f0-4a99-aacc-b372d44f2416.png)

When playing with the `compile_commands.json` file (included in the `.zip`) I noticed that if there was a file without the C++17 flag from the external library (Google Test) included in it, clangd would show this error. If I added the C++17 flag to **all** the files in the JSON then the error would disappear. 

**Logs**
<details>
<summary>Logs</summary>

```
I[08:42:00.753] clangd version 12.0.0 (https://github.com/clangd/llvm-project 8a844a3a321ea56da21af4d48433ad493b1bcfae)
I[08:42:00.755] PID: 12088
I[08:42:00.755] Working directory: c:\cpp-experiments
I[08:42:00.755] argv[0]: c:\Users\jmigual\AppData\Roaming\Code\User\globalStorage\llvm-vs-code-extensions.vscode-clangd\install\12.0.0\clangd_12.0.0\bin\clangd.exe
I[08:42:00.755] argv[1]: --log=verbose
V[08:42:00.763] User config file is C:\Users\jmigual\AppData\Local\clangd\config.yaml
I[08:42:00.763] Starting LSP over stdin/stdout
V[08:42:00.763] <<< {"id":0,"jsonrpc":"2.0","method":"initialize","params":{"capabilities":{"general":{"markdown":{"parser":"marked","version":"1.1.0"},"regularExpressions":{"engine":"ECMAScript","version":"ES2020"},"staleRequestSupport":{"cancel":true,"retryOnContentModified":["textDocument/semanticTokens/full","textDocument/semanticTokens/range","textDocument/semanticTokens/full/delta"]}},"textDocument":{"callHierarchy":{"dynamicRegistration":true},"codeAction":{"codeActionLiteralSupport":{"codeActionKind":{"valueSet":["","quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite","source","source.organizeImports"]}},"dataSupport":true,"disabledSupport":true,"dynamicRegistration":true,"honorsChangeAnnotations":false,"isPreferredSupport":true,"resolveSupport":{"properties":["edit"]}},"codeLens":{"dynamicRegistration":true},"colorProvider":{"dynamicRegistration":true},"completion":{"completionItem":{"commitCharactersSupport":true,"deprecatedSupport":true,"documentationFormat":["markdown","plaintext"],"insertReplaceSupport":true,"insertTextModeSupport":{"valueSet":[1,2]},"labelDetailsSupport":true,"preselectSupport":true,"resolveSupport":{"properties":["documentation","detail","additionalTextEdits"]},"snippetSupport":true,"tagSupport":{"valueSet":[1]}},"completionItemKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25]},"contextSupport":true,"dynamicRegistration":true,"editsNearCursor":true},"declaration":{"dynamicRegistration":true,"linkSupport":true},"definition":{"dynamicRegistration":true,"linkSupport":true},"documentHighlight":{"dynamicRegistration":true},"documentLink":{"dynamicRegistration":true,"tooltipSupport":true},"documentSymbol":{"dynamicRegistration":true,"hierarchicalDocumentSymbolSupport":true,"labelSupport":true,"symbolKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]},"tagSupport":{"valueSet":[1]}},"foldingRange":{"dynamicRegistration":true,"lineFoldingOnly":true,"rangeLimit":5000},"formatting":{"dynamicRegistration":true},"hover":{"contentFormat":["markdown","plaintext"],"dynamicRegistration":true},"implementation":{"dynamicRegistration":true,"linkSupport":true},"linkedEditingRange":{"dynamicRegistration":true},"onTypeFormatting":{"dynamicRegistration":true},"publishDiagnostics":{"codeDescriptionSupport":true,"dataSupport":true,"relatedInformation":true,"tagSupport":{"valueSet":[1,2]},"versionSupport":false},"rangeFormatting":{"dynamicRegistration":true},"references":{"dynamicRegistration":true},"rename":{"dynamicRegistration":true,"honorsChangeAnnotations":true,"prepareSupport":true,"prepareSupportDefaultBehavior":1},"selectionRange":{"dynamicRegistration":true},"semanticHighlightingCapabilities":{"semanticHighlighting":true},"semanticTokens":{"dynamicRegistration":true,"formats":["relative"],"multilineTokenSupport":false,"overlappingTokenSupport":false,"requests":{"full":{"delta":true},"range":true},"tokenModifiers":["declaration","definition","readonly","static","deprecated","abstract","async","modification","documentation","defaultLibrary"],"tokenTypes":["namespace","type","class","enum","interface","struct","typeParameter","parameter","variable","property","enumMember","event","function","method","macro","keyword","modifier","comment","string","number","regexp","operator"]},"signatureHelp":{"contextSupport":true,"dynamicRegistration":true,"signatureInformation":{"activeParameterSupport":true,"documentationFormat":["markdown","plaintext"],"parameterInformation":{"labelOffsetSupport":true}}},"synchronization":{"didSave":true,"dynamicRegistration":true,"willSave":true,"willSaveWaitUntil":true},"typeDefinition":{"dynamicRegistration":true,"linkSupport":true}},"window":{"showDocument":{"support":true},"showMessage":{"messageActionItem":{"additionalPropertiesSupport":true}},"workDoneProgress":true},"workspace":{"applyEdit":true,"codeLens":{"refreshSupport":true},"configuration":true,"didChangeConfiguration":{"dynamicRegistration":true},"didChangeWatchedFiles":{"dynamicRegistration":true},"executeCommand":{"dynamicRegistration":true},"fileOperations":{"didCreate":true,"didDelete":true,"didRename":true,"dynamicRegistration":true,"willCreate":true,"willDelete":true,"willRename":true},"semanticTokens":{"refreshSupport":true},"symbol":{"dynamicRegistration":true,"symbolKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]},"tagSupport":{"valueSet":[1]}},"workspaceEdit":{"changeAnnotationSupport":{"groupsOnLabel":true},"documentChanges":true,"failureHandling":"textOnlyTransactional","normalizesLineEndings":true,"resourceOperations":["create","rename","delete"]},"workspaceFolders":true}},"clientInfo":{"name":"Visual Studio Code","version":"1.58.2"},"initializationOptions":{"clangdFileStatus":true,"fallbackFlags":[]},"locale":"ca","processId":24084,"rootPath":"c:\\cpp-experiments","rootUri":"file:///c%3A/cpp-experiments","trace":"off","workspaceFolders":[{"name":"cpp-experiments","uri":"file:///c%3A/cpp-experiments"}]}}

I[08:42:00.764] <-- initialize(0)
I[08:42:00.764] Client supports legacy semanticHighlights notification and standard semanticTokens request, choosing the latter (no notifications).
I[08:42:01.231] --> reply:initialize(0) 467 ms
V[08:42:01.231] >>> {"id":0,"jsonrpc":"2.0","result":{"capabilities":{"astProvider":true,"callHierarchyProvider":true,"codeActionProvider":{"codeActionKinds":["quickfix","refactor","info"]},"compilationDatabase":{"automaticReload":true},"completionProvider":{"allCommitCharacters":[" ","\t","(",")","[","]","{","}","<",">",":",";",",","+","-","/","*","%","^","&","#","?",".","=","\"","'","|"],"resolveProvider":false,"triggerCharacters":[".","<",">",":","\"","/"]},"declarationProvider":true,"definitionProvider":true,"documentFormattingProvider":true,"documentHighlightProvider":true,"documentLinkProvider":{"resolveProvider":false},"documentOnTypeFormattingProvider":{"firstTriggerCharacter":"\n","moreTriggerCharacter":[]},"documentRangeFormattingProvider":true,"documentSymbolProvider":true,"executeCommandProvider":{"commands":["clangd.applyFix","clangd.applyTweak"]},"hoverProvider":true,"implementationProvider":true,"memoryUsageProvider":true,"referencesProvider":true,"renameProvider":{"prepareProvider":true},"selectionRangeProvider":true,"semanticTokensProvider":{"full":{"delta":true},"legend":{"tokenModifiers":[],"tokenTypes":["variable","variable","parameter","function","method","function","property","variable","class","enum","enumMember","type","dependent","dependent","namespace","typeParameter","concept","type","macro","comment"]},"range":false},"signatureHelpProvider":{"triggerCharacters":["(",","]},"textDocumentSync":{"change":2,"openClose":true,"save":true},"typeHierarchyProvider":true,"workspaceSymbolProvider":true},"serverInfo":{"name":"clangd","version":"clangd version 12.0.0 (https://github.com/clangd/llvm-project 8a844a3a321ea56da21af4d48433ad493b1bcfae)"}}}

V[08:42:01.237] <<< {"jsonrpc":"2.0","method":"initialized","params":{}}

I[08:42:01.237] <-- initialized
V[08:42:01.238] <<< {"jsonrpc":"2.0","method":"textDocument/didOpen","params":{"textDocument":{"languageId":"cpp","text":"//\n// Created by jmarc on 11/6/2021.\n//\n\n#ifndef MATH_INTERVAL_HPP\n#define MATH_INTERVAL_HPP\n\n#include <optional>\n\nnamespace Math {\n\nstruct Interval {\n  const std::optional<int> min, max;\n};\n} // namespace Math\n\n#endif // MATH_INTERVAL_HPP\n","uri":"file:///c%3A/cpp-experiments/include/math/my_header.hpp","version":28}}}

I[08:42:01.238] <-- textDocument/didOpen
I[08:42:01.240] Loaded compilation database from c:\cpp-experiments\compile_commands.json
V[08:42:01.240] Broadcasting compilation database from c:\cpp-experiments
I[08:42:01.248] Loaded compilation database from C:\cpp-experiments\build\VSCode\compile_commands.json
I[08:42:01.249] --> window/workDoneProgress/create(0)
V[08:42:01.249] >>> {"id":0,"jsonrpc":"2.0","method":"window/workDoneProgress/create","params":{"token":"backgroundIndexProgress"}}

V[08:42:01.249] <<< {"id":1,"jsonrpc":"2.0","method":"textDocument/documentSymbol","params":{"textDocument":{"uri":"file:///c%3A/cpp-experiments/include/math/my_header.hpp"}}}

I[08:42:01.249] <-- textDocument/documentSymbol(1)
I[08:42:01.249] Enqueueing 3 commands for indexing
V[08:42:01.250] <<< {"id":2,"jsonrpc":"2.0","method":"textDocument/codeAction","params":{"context":{"diagnostics":[]},"range":{"end":{"character":27,"line":16},"start":{"character":27,"line":16}},"textDocument":{"uri":"file:///c%3A/cpp-experiments/include/math/my_header.hpp"}}}

I[08:42:01.250] <-- textDocument/codeAction(2)
V[08:42:01.250] <<< {"id":3,"jsonrpc":"2.0","method":"textDocument/documentLink","params":{"textDocument":{"uri":"file:///c%3A/cpp-experiments/include/math/my_header.hpp"}}}

I[08:42:01.250] <-- textDocument/documentLink(3)
V[08:42:01.251] <<< {"id":0,"jsonrpc":"2.0","result":null}

I[08:42:01.251] <-- reply(0)
I[08:42:01.251] --> $/progress
V[08:42:01.251] >>> {"jsonrpc":"2.0","method":"$/progress","params":{"token":"backgroundIndexProgress","value":{"kind":"begin","percentage":0,"title":"indexing"}}}

I[08:42:01.251] --> $/progress
V[08:42:01.251] >>> {"jsonrpc":"2.0","method":"$/progress","params":{"token":"backgroundIndexProgress","value":{"kind":"report","message":"0/1","percentage":0}}}

V[08:42:01.548] <<< {"id":4,"jsonrpc":"2.0","method":"textDocument/semanticTokens/full","params":{"textDocument":{"uri":"file:///c%3A/cpp-experiments/include/math/my_header.hpp"}}}

I[08:42:01.548] <-- textDocument/semanticTokens/full(4)
I[08:42:01.562] ASTWorker building file c:\cpp-experiments\include\math\my_header.hpp version 28 with command inferred from C:\cpp-experiments\build\VSCode\_deps\googletest-src\googlemock\src\gmock-all.cc
[C:/cpp-experiments/build/VSCode]
C:\Program Files\LLVM\bin\CLANG_~1.EXE -IC:/cpp-experiments/build/VSCode/_deps/googletest-src/googlemock/include -IC:/cpp-experiments/build/VSCode/_deps/googletest-src/googlemock -isystem C:/cpp-experiments/build/VSCode/_deps/googletest-src/googletest/include -isystem C:/cpp-experiments/build/VSCode/_deps/googletest-src/googletest -g -Xclang -gcodeview -O0 -D_DEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrtd -Wall -Wshadow -Wconversion -DGTEST_HAS_PTHREAD=0 -fexceptions -W -Wpointer-arith -Wreturn-type -Wcast-qual -Wwrite-strings -Wswitch -Wunused-parameter -Wcast-align -Wchar-subscripts -Winline -Wredundant-decls -c c:\cpp-experiments\include\math\my_header.hpp -fsyntax-only -resource-dir=C:\Users\jmigual\AppData\Roaming\Code\User\globalStorage\llvm-vs-code-extensions.vscode-clangd\install\12.0.0\clangd_12.0.0\lib\clang\12.0.0
V[08:42:01.574] Driver produced command: cc1 -cc1 -triple x86_64-pc-windows-msvc19.29.30037 -fsyntax-only -disable-free -disable-llvm-verifier -discard-value-names -main-file-name my_header.hpp -mrelocation-model pic -pic-level 2 -mframe-pointer=none -fmath-errno -fno-rounding-math -mconstructor-aliases -munwind-tables -target-cpu x86-64 -tune-cpu generic -gno-column-info -gcodeview -debug-info-kind=limited -resource-dir C:\Users\jmigual\AppData\Roaming\Code\User\globalStorage\llvm-vs-code-extensions.vscode-clangd\install\12.0.0\clangd_12.0.0\lib\clang\12.0.0 -isystem C:/cpp-experiments/build/VSCode/_deps/googletest-src/googletest/include -isystem C:/cpp-experiments/build/VSCode/_deps/googletest-src/googletest -I C:/cpp-experiments/build/VSCode/_deps/googletest-src/googlemock/include -I C:/cpp-experiments/build/VSCode/_deps/googletest-src/googlemock -D _DEBUG -D _DLL -D _MT -D GTEST_HAS_PTHREAD=0 -internal-isystem C:\Users\jmigual\AppData\Roaming\Code\User\globalStorage\llvm-vs-code-extensions.vscode-clangd\install\12.0.0\clangd_12.0.0\lib\clang\12.0.0\include -internal-isystem C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include -internal-isystem C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\atlmfc\include -internal-isystem C:\Program Files (x86)\Windows Kits\10\Include\10.0.19041.0\ucrt -internal-isystem C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared -internal-isystem C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um -internal-isystem C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\winrt -O0 -Wall -Wshadow -Wconversion -W -Wpointer-arith -Wreturn-type -Wcast-qual -Wwrite-strings -Wswitch -Wunused-parameter -Wcast-align -Wchar-subscripts -Winline -Wredundant-decls -fdeprecated-macro -fdebug-compilation-dir C:/cpp-experiments/build/VSCode -ferror-limit 19 -fno-use-cxa-atexit -fms-extensions -fms-compatibility -fms-compatibility-version=19.29.30037 -std=c++14 -fdelayed-template-parsing -fcxx-exceptions -fexceptions -gcodeview --dependent-lib=msvcrtd -faddrsig -x c++-header c:\cpp-experiments\include\math\my_header.hpp
I[08:42:01.574] --> textDocument/clangd.fileStatus
V[08:42:01.574] >>> {"jsonrpc":"2.0","method":"textDocument/clangd.fileStatus","params":{"state":"parsing includes, running Update (28)","uri":"file:///c:/cpp-experiments/include/math/my_header.hpp"}}

V[08:42:01.574] Building first preamble for c:\cpp-experiments\include\math\my_header.hpp version 28
V[08:42:01.741] indexed preamble AST for c:\cpp-experiments\include\math\my_header.hpp version 28:
  symbol slab: 1699 symbols, 459832 bytes
  ref slab: 0 symbols, 0 refs, 136 bytes
  relations slab: 0 relations, 24 bytes
V[08:42:01.774] Build dynamic index for header symbols with estimated memory usage of 1864484 bytes
V[08:42:01.782] Built preamble of size 607912 for file c:\cpp-experiments\include\math\my_header.hpp version 28
V[08:42:01.802] Trying to fix unresolved name "optional" in scopes: [std::]
V[08:42:01.802] Dex query tree: (LIMIT 10000 (& T=ona S=std:: T=nal T=ion T=tio T=pti T=opt ?=Restricted For Code Completion))
V[08:42:01.803] indexed file AST for c:\cpp-experiments\include\math\my_header.hpp version 28:
  symbol slab: 5 symbols, 5384 bytes
  ref slab: 4 symbols, 4 refs, 4328 bytes
  relations slab: 0 relations, 24 bytes
V[08:42:01.803] Build dynamic index for main-file symbols with estimated memory usage of 13560 bytes
I[08:42:01.803] --> textDocument/publishDiagnostics
V[08:42:01.803] >>> {"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"diagnostics":[{"code":"no_member_template","message":"No template named 'optional' in namespace 'std'","range":{"end":{"character":21,"line":12},"start":{"character":13,"line":12}},"relatedInformation":[],"severity":1,"source":"clang"}],"uri":"file:///c:/cpp-experiments/include/math/my_header.hpp","version":28}}

V[08:42:01.803] ASTWorker running DocumentSymbols on version 28 of c:\cpp-experiments\include\math\my_header.hpp
I[08:42:01.803] --> reply:textDocument/documentSymbol(1) 554 ms
V[08:42:01.803] >>> {"id":1,"jsonrpc":"2.0","result":[{"children":[{"children":[{"kind":8,"name":"min","range":{"end":{"character":30,"line":12},"start":{"character":2,"line":12}},"selectionRange":{"end":{"character":30,"line":12},"start":{"character":27,"line":12}}},{"kind":8,"name":"max","range":{"end":{"character":35,"line":12},"start":{"character":2,"line":12}},"selectionRange":{"end":{"character":35,"line":12},"start":{"character":32,"line":12}}}],"kind":23,"name":"Interval","range":{"end":{"character":1,"line":13},"start":{"character":0,"line":11}},"selectionRange":{"end":{"character":15,"line":11},"start":{"character":7,"line":11}}}],"kind":3,"name":"Math","range":{"end":{"character":1,"line":14},"start":{"character":0,"line":9}},"selectionRange":{"end":{"character":14,"line":9},"start":{"character":10,"line":9}}}]}

V[08:42:01.803] ASTWorker running EnumerateTweaks on version 28 of c:\cpp-experiments\include\math\my_header.hpp
I[08:42:01.803] --> reply:textDocument/codeAction(2) 552 ms
V[08:42:01.803] >>> {"id":2,"jsonrpc":"2.0","result":[]}

V[08:42:01.803] ASTWorker running DocumentLinks on version 28 of c:\cpp-experiments\include\math\my_header.hpp
I[08:42:01.803] --> reply:textDocument/documentLink(3) 552 ms
V[08:42:01.803] >>> {"id":3,"jsonrpc":"2.0","result":[{"range":{"end":{"character":19,"line":7},"start":{"character":9,"line":7}},"target":"file:///C:/Program%20Files%20%28x86%29/Microsoft%20Visual%20Studio/2019/Community/VC/Tools/MSVC/14.29.30037/include/optional"}]}

V[08:42:01.803] ASTWorker running SemanticHighlights on version 28 of c:\cpp-experiments\include\math\my_header.hpp
I[08:42:01.803] --> reply:textDocument/semanticTokens/full(4) 255 ms
V[08:42:01.803] >>> {"id":4,"jsonrpc":"2.0","result":{"data":[4,8,17,18,0,1,8,17,18,0,4,10,4,14,0,2,7,8,8,0,1,27,3,6,0,0,5,3,6,0],"resultId":"1"}}

I[08:42:01.803] --> textDocument/clangd.fileStatus
V[08:42:01.803] >>> {"jsonrpc":"2.0","method":"textDocument/clangd.fileStatus","params":{"state":"idle","uri":"file:///c:/cpp-experiments/include/math/my_header.hpp"}}

V[08:42:02.355] BackgroundIndex: building version 1 after loading index from disk
V[08:42:02.494] BackgroundIndex: serving version 1 (9837431 bytes)
I[08:42:03.127] --> $/progress
V[08:42:03.127] >>> {"jsonrpc":"2.0","method":"$/progress","params":{"token":"backgroundIndexProgress","value":{"kind":"report","message":"0/2","percentage":0}}}

I[08:42:03.127] --> $/progress
V[08:42:03.127] >>> {"jsonrpc":"2.0","method":"$/progress","params":{"token":"backgroundIndexProgress","value":{"kind":"report","message":"0/2","percentage":0}}}

I[08:42:03.127] --> $/progress
V[08:42:03.127] >>> {"jsonrpc":"2.0","method":"$/progress","params":{"token":"backgroundIndexProgress","value":{"kind":"report","message":"1/2","percentage":50}}}

V[08:42:03.128] Indexing C:\cpp-experiments\src\main.cpp (digest:=07BC1D91F6873127)
I[08:42:03.697] Indexed C:\cpp-experiments\src\main.cpp (5 symbols, 16798 refs, 113 files)
V[08:42:03.724] BackgroundIndex: building version 2 when background indexer is idle
V[08:42:03.850] BackgroundIndex: serving version 2 (9831747 bytes)
I[08:42:03.851] --> $/progress
V[08:42:03.851] >>> {"jsonrpc":"2.0","method":"$/progress","params":{"token":"backgroundIndexProgress","value":{"kind":"end"}}}

V[08:42:26.301] <<< {"id":5,"jsonrpc":"2.0","method":"textDocument/documentHighlight","params":{"position":{"character":19,"line":14},"textDocument":{"uri":"file:///c%3A/cpp-experiments/include/math/my_header.hpp"}}}

I[08:42:26.301] <-- textDocument/documentHighlight(5)
V[08:42:26.303] ASTWorker running Highlights on version 28 of c:\cpp-experiments\include\math\my_header.hpp
I[08:42:26.303] --> reply:textDocument/documentHighlight(5) 1 ms
V[08:42:26.303] >>> {"id":5,"jsonrpc":"2.0","result":[]}

I[08:42:26.303] --> textDocument/clangd.fileStatus
V[08:42:26.303] >>> {"jsonrpc":"2.0","method":"textDocument/clangd.fileStatus","params":{"state":"idle","uri":"file:///c:/cpp-experiments/include/math/my_header.hpp"}}

V[08:42:26.503] <<< {"id":6,"jsonrpc":"2.0","method":"textDocument/codeAction","params":{"context":{"diagnostics":[]},"range":{"end":{"character":19,"line":14},"start":{"character":19,"line":14}},"textDocument":{"uri":"file:///c%3A/cpp-experiments/include/math/my_header.hpp"}}}

I[08:42:26.503] <-- textDocument/codeAction(6)
V[08:42:26.504] ASTWorker running EnumerateTweaks on version 28 of c:\cpp-experiments\include\math\my_header.hpp
I[08:42:26.504] --> reply:textDocument/codeAction(6) 0 ms
V[08:42:26.504] >>> {"id":6,"jsonrpc":"2.0","result":[]}

I[08:42:26.504] --> textDocument/clangd.fileStatus
V[08:42:26.504] >>> {"jsonrpc":"2.0","method":"textDocument/clangd.fileStatus","params":{"state":"idle","uri":"file:///c:/cpp-experiments/include/math/my_header.hpp"}}

V[08:42:32.636] <<< {"id":7,"jsonrpc":"2.0","method":"textDocument/hover","params":{"position":{"character":12,"line":11},"textDocument":{"uri":"file:///c%3A/cpp-experiments/include/math/my_header.hpp"}}}

I[08:42:32.636] <-- textDocument/hover(7)
V[08:42:32.637] ASTWorker running Hover on version 28 of c:\cpp-experiments\include\math\my_header.hpp
I[08:42:32.638] --> reply:textDocument/hover(7) 2 ms
V[08:42:32.638] >>> {"id":7,"jsonrpc":"2.0","result":{"contents":{"kind":"markdown","value":"### struct `Interval`  \n\n---\n```cpp\n// In namespace Math\npublic: struct Interval {}\n```"},"range":{"end":{"character":15,"line":11},"start":{"character":7,"line":11}}}}

I[08:42:32.638] --> textDocument/clangd.fileStatus
V[08:42:32.638] >>> {"jsonrpc":"2.0","method":"textDocument/clangd.fileStatus","params":{"state":"idle","uri":"file:///c:/cpp-experiments/include/math/my_header.hpp"}}

V[08:42:32.958] <<< {"id":8,"jsonrpc":"2.0","method":"textDocument/hover","params":{"position":{"character":17,"line":12},"textDocument":{"uri":"file:///c%3A/cpp-experiments/include/math/my_header.hpp"}}}

I[08:42:32.958] <-- textDocument/hover(8)
V[08:42:32.958] ASTWorker running Hover on version 28 of c:\cpp-experiments\include\math\my_header.hpp
I[08:42:32.959] --> reply:textDocument/hover(8) 1 ms
V[08:42:32.959] >>> {"id":8,"jsonrpc":"2.0","result":{"contents":{"kind":"markdown","value":"### field `min`  \n\n---\nType: `const int`  \n\n---\n```cpp\n// In Interval\npublic: const int min\n```"},"range":{"end":{"character":21,"line":12},"start":{"character":13,"line":12}}}}

I[08:42:32.959] --> textDocument/clangd.fileStatus
V[08:42:32.959] >>> {"jsonrpc":"2.0","method":"textDocument/clangd.fileStatus","params":{"state":"idle","uri":"file:///c:/cpp-experiments/include/math/my_header.hpp"}}

V[08:42:33.123] <<< {"id":9,"jsonrpc":"2.0","method":"textDocument/codeAction","params":{"context":{"diagnostics":[{"code":"no_member_template","message":"No template named 'optional' in namespace 'std'","range":{"end":{"character":21,"line":12},"start":{"character":13,"line":12}},"relatedInformation":[],"severity":1,"source":"clang"}],"only":["quickfix"]},"range":{"end":{"character":21,"line":12},"start":{"character":13,"line":12}},"textDocument":{"uri":"file:///c%3A/cpp-experiments/include/math/my_header.hpp"}}}

I[08:42:33.124] <-- textDocument/codeAction(9)
V[08:42:33.124] ASTWorker running EnumerateTweaks on version 28 of c:\cpp-experiments\include\math\my_header.hpp
I[08:42:33.124] --> reply:textDocument/codeAction(9) 0 ms
V[08:42:33.124] >>> {"id":9,"jsonrpc":"2.0","result":[]}

I[08:42:33.124] --> textDocument/clangd.fileStatus
V[08:42:33.124] >>> {"jsonrpc":"2.0","method":"textDocument/clangd.fileStatus","params":{"state":"idle","uri":"file:///c:/cpp-experiments/include/math/my_header.hpp"}}
```

</details>
<!-- Please attach the clangd stderr log if you can. (Usually available from the editor)
If possible, run with `--log=verbose` - note that the logs will include the contents of open files!
If this is a crash, try to put `llvm-symbolizer` on your PATH per the troubleshooting instructions. --!>

**System information**
Output of `clangd --version`: `clangd version 12.0.0`
Editor/LSP plugin: VSCode
Operating system: Windows 10 version 21H1 (19043.1110)

**Comment by HighCommander4** [comment]
The important part of the log is:

> ASTWorker building file c:\cpp-experiments\include\math\my_header.hpp version 28 with command inferred from C:\cpp-experiments\build\VSCode\_deps\googletest-src\googlemock\src\gmock-all.cc

I take it this is the file without the C++17 flag (and which therefore we would prefer not to infer the command from).

Where are the files that do have the C++17 flag located?

**Comment by HighCommander4** [comment]
> Where are the files that do have the C++17 flag located?

To elaborate a bit, clangd uses some heuristics to choose a source file whose flags to use to parse the header file. The heuristics are based mostly on the locations of the files in the directory tree (for example, a source file next to the header file would be preferred).

The bigger issue is using heuristics in the first place. It would be better to look at inclusion relationships to choose a source file that actually includes the header; see #123 for some discussion about that.

**Comment by jmigual** [comment]
Hi @HighCommander4 you are right the `gmock-all.cc` file is the one without the C++17 flag as it is part of an external library.

The file structure in the example is:

```
cpp-experiments/
├── src/
│   ├── main.cpp
│   └── my_library.cpp
├── include/
│   ├── math/
│   │   └── my_header.hpp
│   └── my_library.hpp
└── build/_deps/googletest-src/googlemock/src/
     ├── gmock-all.cc
     └ ...
```

So, all the files inside `src/` have the C++17 flag while the ones inside `build/.../src/` do not have C++17 support. All the headers inside `include/` are included by the files in `src/` so they should be parsed with the C++17 flag. The files in `include/` are never included by the files in `build/.../src/`.

The issue that you linked indeed seems to explain the issue. Should I leave this issue open or should I close it then ? (even if it is not solved)

**Comment by HighCommander4** [comment]
It seems like, at the very least, for `my_library.hpp`, the heuristics should be preferring `my_library.cpp` due to the matching file stem.

Less sure about `my_header.hpp`... I think perhaps having an additional directory grouping `include` and `src` (but not `build`) should get the heuristics to prefer a file in `src`? (Haven't actually tried it though.)

**Comment by HighCommander4** [comment]
Another idea: consider excluding the googlemock files from the CBD if you don't actually edit them?

**Comment by jmigual** [comment]
Yes, `my_library.hpp` seems to be propelry matched. Unfortunately I don't think I can group `include` and `src` as this is a shared project (I'd love to do that and it's an open ticket). 

To remove from the CDB you mean from the `compile_commands.json` file? If so, this is a generated file from CMake so even if I manually deleted the lines, as soon as VSCode performs another configure (every time the editor starts) the file would be replaced by the generated one. Is there a clang flag that lets me exclude some files?

**Comment by HighCommander4** [comment]
> To remove from the CDB you mean from the `compile_commands.json` file?

Yes, sorry, "CDB" = "compilation database" = `compile_commands.json`

> If so, this is a generated file from CMake so even if I manually deleted the lines, as soon as VSCode performs another configure (every time the editor starts) the file would be replaced by the generated one. Is there a clang flag that lets me exclude some files?

Clangd has a way to skip **indexing** of some paths in the CDB, but I don't think that's going to affect the command interpolation.

I sometimes post-process my CDB with a JSON processing tool like `jq`, e.g.:

`jq 'map(select(.file | contains("googlemock")))' compile_commands.json | sponge compile_commands.json`

Not sure off the top of my head how to integrate that into a cmake workflow.

(To be clear, these are suggestions for workarounds. I think a proper fix would be a more comprehensive fix for #123 on the clangd side.)

**Comment by jmigual** [comment]
I followed your workaround and for now it seems to work. Thank you for taking your time!
